lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1395,19 +1395,23 @@ bool TypeChecker::coercePatternToType(Pattern *&P, TypeResolution resolution,
             goto recur;
           }
 
-          auto diag = diagnose(EEP->getLoc(),
-                               diag::enum_element_pattern_member_not_found,
-                               EEP->getName().str(), type);
-
-          // If we have an optional type let's try to see if the case
-          // exists in its base type, if so we can suggest a fix-it for that.
+          // If we have an optional type, let's try to see if the case
+          // exists in its base type and if it does then synthesize an
+          // OptionalSomePattern that wraps the case.
           if (auto baseType = type->getOptionalObjectType()) {
```

Comment:
This presumably ought to work recursively.  You can just count how many optional types you looked through and then wrap it in that many `OptionalSomePattern`s.

Reply:
Would be great to add a test for that as well (if we don't have one yet).

Reply:
Ah right, looking through all the optionals instead of only one does the trick. I have added some tests for it as well. @rjmccall @xedin